### PR TITLE
Return null from Layer.find when transform layer has a non-invertable transform

### LINKED ISSUE #18847: Transforming ListView to x scale of 0 throws an exception
If you attempt to transform a ListView to an x scale of 0.0, an error is thrown and nothing will draw.  This is currently something that we do in our app to create a flipping effect, and it no longer works.   
  
The issue seems that some rendering attempts to invert the matrix and the matrix is not able to be inverted.

## Steps to Reproduce

1. flutter create a new project and use the following code:
```
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final Matrix4 transform = Matrix4.identity()..scale(0.0, 1.0, 1.0);
    return new MaterialApp(
      home: new Transform(
        transform: transform,
        child: new ListView(),
      ),
    );
  }
}
```

## Logs

```
flutter: ══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
flutter: The following ArgumentError was thrown during a scheduler callback:
flutter: Invalid argument (other): Matrix cannot be inverted: Instance of 'Matrix4'
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #0      new Matrix4.inverted (file:///Users/KG/.pub-cache/hosted/pub.dartlang.org/vector_math-2.0.6/lib/src/vector_math_64/matrix4.dart:237:7)
flutter: #1      TransformLayer.find (package:flutter/src/rendering/layer.dart:780:32)
flutter: #2      ContainerLayer.find (package:flutter/src/rendering/layer.dart:345:36)
flutter: #3      OffsetLayer.find (package:flutter/src/rendering/layer.dart:546:18)
flutter: #4      ContainerLayer.find (package:flutter/src/rendering/layer.dart:345:36)
flutter: #5      OffsetLayer.find (package:flutter/src/rendering/layer.dart:546:18)
flutter: #6      ContainerLayer.find (package:flutter/src/rendering/layer.dart:345:36)
flutter: #7      OffsetLayer.find (package:flutter/src/rendering/layer.dart:546:18)
flutter: #8      TransformLayer.find (package:flutter/src/rendering/layer.dart:783:18)
flutter: #9      RenderView._updateSystemChrome (package:flutter/src/rendering/view.dart:215:58)
flutter: #10     RenderView.compositeFrame (package:flutter/src/rendering/view.dart:198:9)
flutter: #11     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:273:16)
flutter: #12     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:627:13)
flutter: #13     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:208:5)
flutter: #14     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:990:15)
flutter: #15     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:930:9)
flutter: #16     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.scheduleWarmUpFrame.<anonymous closure> (package:flutter/src/scheduler/binding.dart:751:7)
flutter: #18     _Timer._runTimers (dart:isolate/runtime/libtimer_impl.dart:382:19)
flutter: #19     _Timer._handleMessage (dart:isolate/runtime/libtimer_impl.dart:416:5)
flutter: #20     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:165:12)
flutter: (elided one frame from package dart:async)
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
flutter: Another exception was thrown: Invalid argument (other): Matrix cannot be inverted: Instance of 'Matrix4'
```

```
Analyzing matrixtests...                                         
No issues found! (ran in 1.6s)
```

```
[✓] Flutter (Channel master, v0.5.6-pre.76, on Mac OS X 10.13.5 17F77, locale en-US)
    • Flutter version 0.5.6-pre.76 at /Users/KG/Developer/Flutter/flutter
    • Framework revision 579116412f (10 hours ago), 2018-06-26 12:18:50 +0200
    • Engine revision 549c855e89
    • Dart version 2.0.0-dev.63.0.flutter-4c9689c1d2

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/KG/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = /Users/KG/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.4, Build version 9F1027a
    • ios-deploy 1.9.2
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 25.0.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] IntelliJ IDEA Community Edition (version 2017.2.7)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 21.2.2
    • Dart plugin version 172.4343.25

[!] VS Code (version 1.21.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected devices (1 available)
    • iPhone • 69f9d7937b3ddbd746dfed588abfb982f5dd6811 • ios • iOS 11.3
```

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/18847

### CODE REVIEW

File: packages/flutter/lib/src/rendering/layer.dart [hunk 1]
```diff
@@ -757,10 +757,12 @@ class TransformLayer extends OffsetLayer {
       return;
     _transform = value;
     _invertedTransform = null;
+    _lastDeterminate = null;
   }
 
   Matrix4 _lastEffectiveTransform;
   Matrix4 _invertedTransform;
+  double _lastDeterminate;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/layer.dart]: i think you mean determinant ?

Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/layer.dart]: D'oh

**Comment by Hixie** [comment]
Rather than tracking the determinant, another option would be to just have a "dirty" flag, which controls the computation of the inverse, and then (regardless of dirtiness) bail if the matrix is null.

Either is fine. LGTM.

File: packages/flutter/lib/src/rendering/layer.dart [hunk 2]
```diff
@@ -1187,8 +1197,7 @@ class FollowerLayer extends ContainerLayer {
     if (_invertedTransform == null) {
       final Matrix4 transform = getLastTransform();
       assert(transform != null);
-      _invertedTransform = new Matrix4.zero();
-      transform.copyInverse(_invertedTransform);
+      _invertedTransform = new Matrix4.inverted(transform);
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/layer.dart]: In my head, a translation matrix shouldn't ever be singular.

**Comment by jonahwilliams** [comment]
Updated to flag, added check to both layers since Link could still have a transformation apart from the translation

**Pull request merged**